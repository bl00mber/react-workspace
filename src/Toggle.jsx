class Toggle extends React.Component {
  constructor(props) {
    super(props);
    this.state = { isToggleOn: false };
  }

  handleToggle(e, toggleValue) {
    this.setState(prevState => ({
      isToggleOn: toggleValue !== undefined ? toggleValue : !prevState.isToggleOn
    }));
  }

  render() {
    return (
      <div>
        <p>Is toggled {this.state.isToggleOn ? 'ON' : 'OFF'}</p>
        <button onClick={(e) => this.handleToggle(e)}>Toggle</button>
        <button onClick={(e) => this.handleToggle(e, true)}>Toggle ON</button>
        <button onClick={(e) => this.handleToggle(e, false)}>Toggle OFF</button>
      </div>
    );
  }
}
