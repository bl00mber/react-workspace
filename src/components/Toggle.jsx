import React from 'react';
import PropTypes from 'prop-types';

const Toggle = ({ isToggleOn, onToggle }) => (
  <div>
    <p>
      Is toggled
      {isToggleOn ? 'ON' : 'OFF'}
    </p>
    <button type="button" onClick={() => onToggle()}>Toggle</button>
    <button type="button" onClick={() => onToggle(true)}>Toggle ON</button>
    <button type="button" onClick={() => onToggle(false)}>Toggle OFF</button>
  </div>
);

Toggle.propTypes = {
  isToggleOn: PropTypes.bool.isRequired,
  onToggle: PropTypes.func.isRequired,
};

export default Toggle;
