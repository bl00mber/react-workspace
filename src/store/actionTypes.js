/*
 * action types
 */

export const ADD_TODO = 'ADD_TODO';
export const TOGGLE_TODO = 'TOGGLE_TODO';
export const SET_VISIBILITY_FILTER = 'SET_VISIBILITY_FILTER';
export const INCREMENT_COUNTER = 'INCREMENT_COUNTER';
export const DECREMENT_COUNTER = 'DECREMENT_COUNTER';
export const TOGGLE_FLAG = 'TOGGLE_FLAG';
export const SET_CELSIUS = 'SET_CELSIUS';
export const SET_FAHRENHEIT = 'SET_FAHRENHEIT';
export const FETCH_POSTS_REQUEST = 'FETCH_POSTS_REQUEST';
export const FETCH_POSTS_FAILURE = 'FETCH_POSTS_FAILURE';
export const FETCH_POSTS_SUCCESS = 'FETCH_POSTS_SUCCESS';
