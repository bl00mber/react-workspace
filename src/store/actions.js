import {
  ADD_TODO,
  TOGGLE_TODO,
  SET_VISIBILITY_FILTER,
  TOGGLE_FLAG,
  SET_CELSIUS,
  SET_FAHRENHEIT,
} from './actionTypes';

/*
 * other constants
 */

export const VisibilityFilters = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_COMPLETED: 'SHOW_COMPLETED',
  SHOW_ACTIVE: 'SHOW_ACTIVE',
};

/* --------------- */


export const addTodo = (id, text) => ({
  type: ADD_TODO,
  id,
  text,
});


export const toggleTodo = id => ({
  type: TOGGLE_TODO,
  id,
});


export const setVisibilityFilter = filter => ({
  type: SET_VISIBILITY_FILTER,
  filter,
});

export const toggleFlag = flag => ({
  type: TOGGLE_FLAG,
  flag,
});

export const setCelsius = temperature => ({
  type: SET_CELSIUS,
  temperature,
});

export const setFahrenheit = temperature => ({
  type: SET_FAHRENHEIT,
  temperature,
});
