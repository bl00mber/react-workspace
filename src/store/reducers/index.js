import { combineReducers } from 'redux';

import counter from './counter';
import todos from './todos';
import visibilityFilter from './visibilityFilter';
import toggle from './toggle';

export default combineReducers({
  counter,
  todos,
  visibilityFilter,
  toggle,
});
