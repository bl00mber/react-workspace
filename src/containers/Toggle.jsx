import { connect } from 'react-redux';

import ToggleComponent from '@/components/Toggle';
import { toggleFlag } from '@/store/actions';

const mapStateToProps = state => ({
  isToggleOn: state.toggle,
});

const mapDispatchToProps = dispatch => ({
  onToggle: flag => dispatch(toggleFlag(flag)),
});

const Toggle = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ToggleComponent);

export default Toggle;
