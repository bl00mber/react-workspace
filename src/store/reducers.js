import { combineReducers } from 'redux';

import { VisibilityFilters } from './actions';
import {
  ADD_TODO,
  TOGGLE_TODO,
  SET_VISIBILITY_FILTER,
  INCREMENT_COUNTER,
  DECREMENT_COUNTER,
} from './actionTypes';

const { SHOW_ALL } = VisibilityFilters;

const counter = (state = 0, action) => {
  switch (action.type) {
    case INCREMENT_COUNTER:
      return state + 1;
    case DECREMENT_COUNTER:
      return state - 1;
    default:
      return state;
  }
};

const todos = (state = [], action) => {
  switch (action.type) {
    case ADD_TODO:
      return [
        ...state,
        {
          text: action.text,
          completed: false,
        },
      ];
    case TOGGLE_TODO:
      return state.map((todo, index) => {
        if (index === action.index) {
          return Object.assign({}, todo, {
            completed: !todo.completed,
          });
        }
        return todo;
      });
    default:
      return state;
  }
};

const visibilityFilter = (state = SHOW_ALL, action) => {
  switch (action.type) {
    case SET_VISIBILITY_FILTER:
      return action.filter;
    default:
      return state;
  }
};

export default combineReducers({
  counter,
  todos,
  visibilityFilter,
});
