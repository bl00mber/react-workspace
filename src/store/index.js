import { createStore } from 'redux';

import actions from './actions';
import actionTypes from './actionTypes';
import reducers from './reducers';

/**
 * This is a reducer, a pure function with (state, action) => state signature.
 * It describes how an action transforms the state into the next state.
 *
 * The shape of the state is up to you: it can be a primitive, an array, an object,
 * or even an Immutable.js data structure. The only important part is that you should
 * not mutate the state object, but return a new object if the state changes.
 *
 * In this example, we use a `switch` statement and strings, but you can use a helper that
 * follows a different convention (such as function maps) if it makes sense for your
 * project.
 */
const counter = (state = 0, action) => {
  switch (action.type) {
    case 'INCREMENT':
      return state + 1;
    case 'DECREMENT':
      return state - 1;
    default:
      return state;
  }
};

// Create a Redux store holding the state of your app.
// Its API is { subscribe, dispatch, getState }.
const isDev = process.env.NODE_ENV === 'development';
const hasReduxDevTools = window.__REDUX_DEVTOOLS_EXTENSION__; // eslint-disable-line
const reduxDevTools = hasReduxDevTools && window.__REDUX_DEVTOOLS_EXTENSION__()  // eslint-disable-line
const store = createStore(
  counter,
  hasReduxDevTools && isDev && reduxDevTools,
);

// You can use subscribe() to update the UI in response to state changes.
// Normally you'd use a view binding library (e.g. React Redux) rather than subscribe() directly.
// However it can also be handy to persist the current state in the localStorage.
// store.subscribe(() => console.log(store.getState()));

export {
  actions,
  actionTypes,
  reducers,
  store,
};

export default store;
