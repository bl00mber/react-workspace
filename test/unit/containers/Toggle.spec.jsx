import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import ToggleComponent from '@/components/Toggle';
import Toggle from '@/containers/Toggle';
import { toggleFlag } from '@/store/actions';

jest.mock('@/store/actions', () => ({
  toggleFlag: jest.fn(),
}));

const state = {
  toggle: false,
};

let store;

beforeEach(() => {
  store = {
    getState: jest.fn().mockReturnValue(state),
    subscribe: jest.fn(),
    dispatch: jest.fn(),
  };

  toggleFlag.mockReset();
  toggleFlag.mockReturnValue('return toggleFlag action');
});

describe('Toggle', () => {
  test('should match enzyme shallow snapshot', () => {
    const props = {
      store,
    };
    const wrapper = shallow(<Toggle {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should have correct props mapped', () => {
    const props = {
      store,
    };
    const wrapper = shallow(<Toggle {...props} />);
    const {
      isToggleOn, onToggle,
    } = wrapper.find(ToggleComponent).props();
    expect(isToggleOn).toBe(state.toggle);
    expect(onToggle).toBeDefined();
    expect(typeof onToggle).toBe('function');
  });

  test('should dispatch in the right way', () => {
    const props = {
      store,
    };
    const wrapper = shallow(<Toggle {...props} />);
    const {
      onToggle,
    } = wrapper.find(ToggleComponent).props();

    onToggle('flag');
    expect(toggleFlag).toHaveBeenCalledWith('flag');
    expect(store.dispatch).toHaveBeenCalledWith('return toggleFlag action');
  });
});
