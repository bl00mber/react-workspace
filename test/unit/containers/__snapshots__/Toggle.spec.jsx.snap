// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle should match enzyme shallow snapshot 1`] = `
<Toggle
  isToggleOn={false}
  onToggle={[Function]}
  store={
    Object {
      "dispatch": [MockFunction],
      "getState": [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": Object {
              "toggle": false,
            },
          },
          Object {
            "isThrow": false,
            "value": Object {
              "toggle": false,
            },
          },
        ],
      },
      "subscribe": [MockFunction] {
        "calls": Array [
          Array [
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [MockFunction],
        "getState": [MockFunction] {
          "calls": Array [
            Array [],
            Array [],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": Object {
                "toggle": false,
              },
            },
            Object {
              "isThrow": false,
              "value": Object {
                "toggle": false,
              },
            },
          ],
        },
        "subscribe": [MockFunction] {
          "calls": Array [
            Array [
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
      },
      "unsubscribe": undefined,
    }
  }
/>
`;
